services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.userprofile.utils;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.userprofile.validation.UserUpdateEvent;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
+ */
+public class UserProfileUpdateHelper {
+
+    public static void update(UserUpdateEvent userUpdateEvent, KeycloakSession session, UserModel currentUser, StoredUserProfile updatedUser) {
+        update(userUpdateEvent, session, currentUser, updatedUser, true);
+    }
+
+
+    public static void update(UserUpdateEvent userUpdateEvent, KeycloakSession session, UserModel currentUser, StoredUserProfile updatedUser, boolean removeMissingAttributes) {
+        RealmModel realm = session.getContext().getRealm();
+
+        if (updatedUser.getAttributes() == null || updatedUser.getAttributes().size() == 0)
+            return;
+
+        //The Idp review does not respect "isEditUserNameAllowed" therefore we have to miss the check here
+        if (!userUpdateEvent.equals(UserUpdateEvent.IdpReview)) {
+            //This step has to be done before email is assigned to the username if isRegistrationEmailAsUsername is set
+            //Otherwise email change will not reflect in username changes.
+            if (updatedUser.getFirstAttribute(UserModel.USERNAME) != null && !realm.isEditUsernameAllowed()) {
+                updatedUser.removeAttribute(UserModel.USERNAME);
+            }
+        }
+
+        if (updatedUser.getFirstAttribute(UserModel.EMAIL) != null && updatedUser.getFirstAttribute(UserModel.EMAIL).isEmpty()) {
+            updatedUser.removeAttribute(UserModel.EMAIL);
+            updatedUser.setAttribute(UserModel.EMAIL, Collections.singletonList(null));
```

Comment:
Why create an attribute with a list of `null` values ?

Reply:
because for example this test expects that the handling of email is "" -> null while this is not the case anywhere else... 
https://github.com/keycloak/keycloak/blob/74988a3f21c26764e01956ff8c9ef5cea865cb2d/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java#L836

Reply:
I invested time in keeping the api of the validation as consistent as possible. But there are inconsistencies in the code 

Thats why I initially put the user update method into the user profile. That would have enabled to override the behaviour for the old very inconsistent code path while having a clean new store method in the future. 

But Stian argued against this as the profile should not care about storage but validation only. So I moved the code to this helper class and out of the SPI interface.